### PR TITLE
Reset guard for blocks ending in clobber

### LINKED ISSUE #20: Missing indirect packet guards
Under some conditions we don't emit a packet guard.

`requiredGuard func(insns []instruction) (int, packetGuard)` determines what packet guards a block needs, and how many instructions of `insns` this guard would cover: indirect guards are invalidated by writing to RegX.

But this doesn't allow us to differentiate between a guard that covers all of `insns` as is still valid afterwards and a guard that also covers all of `insns` but is not valid afterwards (because the last instruction in `insns` writes to RegX). This would lead us to think a guard is still valid when it isn't.

Found by fuzzing.

### PR DESCRIPTION
**Pull request by arthurfabre**
requiredGuard func(insns []instruction) (int, packetGuard) determines
what packet guards a block needs, and how many instructions of insns
this guard would cover: indirect guards are invalidated by writing to
RegX.

But this doesn't allow us to differentiate between a guard that covers
all of insns as is still valid afterwards and a guard that also covers
all of insns but is not valid afterwards (because the last instruction
in insns writes to RegX). This would lead us to think a guard is still
valid when it isn't.

Rework requiredGuard() to return 0 for "guard covers all insns and is
still valid". Returning len(insns) now means the guard covers all insns
but isn't valid anymore.

Fixes #20. Found by fuzzing.

### CODE REVIEW

File: cbpfc.go [hunk 1]
```diff
@@ -470,8 +471,9 @@ type packetGuardOpts struct {
 	// requiredGuard returns:
 	// - the packetGuard needed by insns
 	// - the number of instructions in insns covered by the guard.
-	//   The guard is assumed to be invalidated for the remaining / uncovered insns (eg RegX was clobbered for indirect guards).
+	//   The guard is assumed to be invalidated for the next insns (eg RegX was clobbered for indirect guards).
```
Comment (lmb) [inline cbpfc.go:474]: Why `invalidated` not `valid`?

File: cbpfc.go [hunk 2]
```diff
@@ -470,8 +471,9 @@ type packetGuardOpts struct {
...
 	//   requiredGuard will be called until all instructions are covered.
+	//   0 for all instructions covered.
 	requiredGuard func(insns []instruction) (int, packetGuard)
```
Comment (lmb) [inline cbpfc.go:477]: I think I would find a `bool` return value easier to understand. The `0` is kind of clever but imo just leads to subtle bugs.

File: cbpfc.go (hunk 1 above)
Comment (arthurfabre) [reply cbpfc.go:474]: If the returned number of insns was `!= 0` that meant the guard wasn't valid for the next instructions. Reworked the comments to be clearer hopefully.

File: cbpfc.go (hunk 2 above)
Comment (arthurfabre) [reply cbpfc.go:477]: Good point. The triple return value didn't seem very clear so I added a new `requriedGuard` struct that lets us document what the different fields are.

File: cbpfc.go [hunk 3]
```diff
@@ -467,17 +477,25 @@ func addIndirectPacketGuards(blocks []*block) {
...
-	requiredGuard func(insns []instruction) (int, packetGuard)
+	// requiredGuard returns the packetGuard needed by insns, and what insns it is valid for.
+	requiredGuard func(insns []instruction) requiredGuard
 
 	// createInsn creates an instruction that checks the packet length against the guard
 	createInsn func(guard packetGuard) bpf.Instruction
 }
 
+type requiredGuard struct {
+	guard packetGuard
```
Comment (lmb) [inline cbpfc.go:488]: Nit: maybe `offset`?

File: cbpfc.go [hunk 4]
```diff
@@ -467,17 +477,25 @@ func addIndirectPacketGuards(blocks []*block) {
...
+
+	// The guard covers all the requested instructions,
+	// and is still valid afterwards.
+	alwaysValid bool
+
+	// The guard covers n instructions,
+	// and isn't valid for the subsequent n+1: instructions (eg RegX was clobbered for indirect guards).
```
Comment (lmb) [inline cbpfc.go:495]: For my understanding: does the clobber always happen on the last insn?

**Review by lmb — APPROVED** [review]

File: cbpfc.go (hunk 3 above)
Comment (arthurfabre) [reply cbpfc.go:488]: That might have been a better name but we call it `guard` in a bunch of places already :(

File: cbpfc.go (hunk 4 above)
Comment (arthurfabre) [reply cbpfc.go:495]: Yup. That invalidates the guard, and we call `requiredGuard` again for the remaining instructions to see if they need one.

**Pull request merged**